File: Deep/Deep/Features/Onboarding/Transition/OnboardingContentDrift.swift
import SwiftUI

/// The onboarding hand-off happens in two layers: the coordinator crossfades
/// whole screens (pure opacity, so fixed elements — chrome, footer CTAs — hold
/// perfectly still), while each screen's *content* additionally drifts and
/// softens through a `SoftDriftTransition` with its own fade disabled, so the
/// two layers never compound opacities. The drift direction follows the
/// coordinator's routing (forward rises in from below, back sinks out the way
/// it came), read from the `onboardingNavDirection` environment value.
/// Reduce Motion drops the drift and blur (the transition degrades to
/// identity), leaving just the coordinator's crossfade.
private struct OnboardingContentDrift: ViewModifier {
  @Environment(\.onboardingNavDirection) private var direction

  func body(content: Content) -> some View {
    content.transition(SoftDriftTransition(direction: direction, fades: false))
  }
}

extension View {
  /// Apply to a screen's content block (never its footer CTA) so it drifts
  /// gently into place while the coordinator crossfades screens.
  func onboardingContentDrift() -> some View {
    modifier(OnboardingContentDrift())
  }
}

#Preview("Content drift") {
  @Previewable @State var showsFirst = true

  ZStack {
    // Stand-ins for two onboarding screens — no real dependencies. The footer
    // button stays put; only the content block drifts.
    VStack(spacing: .rhythm) {
      Group {
        if showsFirst {
          LinearGradient(colors: [.peachCloud, .blushPowder], startPoint: .top, endPoint: .bottom)
            .clipShape(RoundedRectangle(cornerRadius: .card, style: .continuous))
            .overlay {
              Text("First screen")
                .font(DeepType.sectionTitle)
                .foregroundStyle(.deepPlum)
            }
            .id("first")
        } else {
          LinearGradient(colors: [.skyWash, .moonCream], startPoint: .top, endPoint: .bottom)
            .clipShape(RoundedRectangle(cornerRadius: .card, style: .continuous))
            .overlay {
              Text("Second screen")
                .font(DeepType.sectionTitle)
                .foregroundStyle(.deepPlum)
            }
            .id("second")
        }
      }
      .onboardingContentDrift()

      Button("Toggle") {
        withAnimation(.hush) { showsFirst.toggle() }
      }
    }
    .padding(.edge)
  }
}

File: android/app/src/main/java/io/appbeyond/freelance/deep/feature/onboarding/components/OnboardingContentDrift.kt
package io.appbeyond.freelance.deep.feature.onboarding.components

import androidx.compose.runtime.Composable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier

/**
 * The drift a screen's *content block* wears while the coordinator crossfades
 * whole screens.
 *
 * Ported from Deep/Deep/Features/Onboarding/Transition/OnboardingContentDrift.swift
 * and `SoftDriftTransition.swift`. The hand-off happens in two layers: the
 * coordinator fades whole screens (pure opacity, so fixed elements — the chrome,
 * footer CTAs — hold perfectly still), while each screen's content additionally
 * drifts `Dp.drift` in the routing direction with no fade of its own, so the two
 * layers never compound opacities. The iOS `SoftDrift.veil` blur is dropped:
 * `Modifier.blur` is a no-op below API 31, and blurring anything over the
 * atmosphere turns its orbs into hard discs.
 *
 * The coordinator builds the modifier inside its `AnimatedContent` (it needs
 * that scope's `animateEnterExit`) and provides it here — the Compose twin of
 * iOS injecting `onboardingNavDirection` through the environment. The default is
 * identity, so a leaf screen previews without a coordinator.
 */
val LocalOnboardingContentDrift = staticCompositionLocalOf<Modifier> { Modifier }

/** Apply to a screen's content block — never its footer CTA — so it drifts into place. */
@Composable
fun Modifier.onboardingContentDrift(): Modifier = this.then(LocalOnboardingContentDrift.current)
